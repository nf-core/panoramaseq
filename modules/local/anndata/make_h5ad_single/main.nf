process tsv_to_h5ad_single {
    tag "$meta.id"
    label 'process_single'

    conda "${moduleDir}/environment.yml"
    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
        'oras://community.wave.seqera.io/library/anndata:0.10.9--d13580e4b297da7c':
        'community.wave.seqera.io/library/anndata:0.10.9--1eab54e300e1e584' }"

    input:
    tuple val(meta), path(count_tsv), path(coords_csv)

    output:
    tuple val(meta), path("*.h5ad"), emit: h5ad
    path "versions.yml"            , emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def args = task.ext.args ?: ''
    def prefix = task.ext.prefix ?: "${meta.id}"
    def sample_name_arg = meta.id ? "--sample-name ${meta.id}" : ""
    """
    tsv_to_h5ad_single.py \\
        ${count_tsv} \\
        --coords ${coords_csv} \\
        --output ${prefix}.h5ad \\
        ${sample_name_arg} \\
        ${args}

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        python: \$(python3 -c 'import platform; print(platform.python_version())')
        anndata: \$(python3 -c 'import anndata as ad; print(ad.__version__)')
        pandas: \$(python3 -c 'import pandas as pd; print(pd.__version__)')
        scipy: \$(python3 -c 'import scipy; print(scipy.__version__)')
    END_VERSIONS
    """

    stub:
    def prefix = task.ext.prefix ?: "${meta.id}"
    """
    touch ${prefix}.h5ad

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        python: \$(python3 -c 'import platform; print(platform.python_version())')
        anndata: \$(python3 -c 'import anndata as ad; print(ad.__version__)')
        pandas: \$(python3 -c 'import pandas as pd; print(pd.__version__)')
        scipy: \$(python3 -c 'import scipy; print(scipy.__version__)')
    END_VERSIONS
    """
}
