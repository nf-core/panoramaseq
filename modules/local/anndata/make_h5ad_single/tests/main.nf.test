nextflow_process {

    name "Test Process tsv_to_h5ad_single"
    script "../main.nf"
    process "tsv_to_h5ad_single"
    tag "modules"
    tag "modules_local"
    tag "tsv_to_h5ad_single"
    tag "make_h5ad_single"

    test("tsv_to_h5ad_single - stub") {

        options "-stub"

        when {
            params {
                outdir = "$outputDir"
                panoramaseq_testdata_base_path = "https://raw.githubusercontent.com/francops1722/test-datasets/panoramaseq/testdata/"
            }
            process {
                """
                input[0] = [
                    [ id: 'S09_single' ],
                    file(params.panoramaseq_testdata_base_path + 'counts/S09_counts.tsv.gz', checkIfExists: true),
                    file(params.panoramaseq_testdata_base_path + 'barcodes_coords.csv', checkIfExists: true)
                ]
                """
            }
        }

        then {
            assertAll (
                { assert process.success },
                { assert process.out.h5ad != null },
                { assert process.out.h5ad.get(0).get(1) ==~ ".*\\.h5ad" },
                { assert process.out.versions != null },
                { assert snapshot(process.out.versions).match() }
            )
        }
    }

    test("tsv_to_h5ad_single - S09 sample") {

        when {
            params {
                outdir = "$outputDir"
                panoramaseq_testdata_base_path = "https://raw.githubusercontent.com/francops1722/test-datasets/panoramaseq/testdata/"
            }
            process {
                """
                input[0] = [
                    [ id: 'S09' ],
                    file(params.panoramaseq_testdata_base_path + 'counts/S09_counts.tsv.gz', checkIfExists: true),
                    file(params.panoramaseq_testdata_base_path + 'barcodes_coords.csv', checkIfExists: true)
                ]
                """
            }
        }

        then {
            assertAll (
                { assert process.success },
                { assert process.out.h5ad != null },
                { assert process.out.h5ad.get(0).get(1) ==~ ".*\\.h5ad" },
                { assert file(process.out.h5ad.get(0).get(1)).exists() },
                { assert process.out.versions != null },
                { assert snapshot(process.out.h5ad, process.out.versions).match("S09_sample") }
            )
        }
    }

    test("tsv_to_h5ad_single - S10 sample") {

        when {
            params {
                outdir = "$outputDir"
                panoramaseq_testdata_base_path = "https://raw.githubusercontent.com/francops1722/test-datasets/panoramaseq/testdata/"
            }
            process {
                """
                input[0] = [
                    [ id: 'S10' ],
                    file(params.panoramaseq_testdata_base_path + 'counts/S10_counts.tsv.gz', checkIfExists: true),
                    file(params.panoramaseq_testdata_base_path + 'barcodes_coords.csv', checkIfExists: true)
                ]
                """
            }
        }

        then {
            assertAll (
                { assert process.success },
                { assert process.out.h5ad != null },
                { assert process.out.h5ad.get(0).get(1) ==~ ".*\\.h5ad" },
                { assert file(process.out.h5ad.get(0).get(1)).exists() },
                { assert process.out.versions != null },
                { assert snapshot(process.out.h5ad, process.out.versions).match("S10_sample") }
            )
        }
    }

    test("tsv_to_h5ad_single - custom prefix") {

        when {
            params {
                outdir = "$outputDir"
                panoramaseq_testdata_base_path = "https://raw.githubusercontent.com/francops1722/test-datasets/panoramaseq/testdata/"
            }
            process {
                """
                input[0] = [
                    [ id: 'custom_test' ],
                    file(params.panoramaseq_testdata_base_path + 'counts/S09_counts.tsv.gz', checkIfExists: true),
                    file(params.panoramaseq_testdata_base_path + 'barcodes_coords.csv', checkIfExists: true)
                ]
                """
            }
        }

        then {
            assertAll (
                { assert process.success },
                { assert process.out.h5ad != null },
                { assert process.out.h5ad.get(0).get(1) ==~ ".*custom_test\\.h5ad" },
                { assert file(process.out.h5ad.get(0).get(1)).exists() },
                { assert process.out.versions != null },
                { assert snapshot(process.out.h5ad, process.out.versions).match("custom_prefix") }
            )
        }
    }
}
