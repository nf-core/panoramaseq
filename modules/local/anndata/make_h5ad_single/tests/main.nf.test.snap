{
    "custom_prefix": {
        "content": [
            [
                [
                    {
                        "id": "custom_test"
                    },
                    "custom_test.h5ad:md5,dc890a21299938872612b8c59a8bd22d"
                ]
            ],
            [
                "versions.yml:md5,7e3208653f573e1a481b9d8597ccc89e"
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.2"
        },
        "timestamp": "2025-07-02T10:50:57.451360719"
    },
    "S09_sample": {
        "content": [
            [
                [
                    {
                        "id": "S09"
                    },
                    "S09.h5ad:md5,a112afc616b277bbc1410b31e2e31a3a"
                ]
            ],
            [
                "versions.yml:md5,7e3208653f573e1a481b9d8597ccc89e"
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.2"
        },
        "timestamp": "2025-07-02T10:50:42.119431078"
    },
    "tsv_to_h5ad_single - stub": {
        "content": [
            [
                "versions.yml:md5,7e3208653f573e1a481b9d8597ccc89e"
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.2"
        },
        "timestamp": "2025-07-02T10:50:34.484590348"
    },
    "S10_sample": {
        "content": [
            [
                [
                    {
                        "id": "S10"
                    },
                    "S10.h5ad:md5,5524b7e2a395dd408df60caee7e1bc61"
                ]
            ],
            [
                "versions.yml:md5,7e3208653f573e1a481b9d8597ccc89e"
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.2"
        },
        "timestamp": "2025-07-02T10:50:49.677500891"
    }
}